### ISSUE TITLE
Add unique index tracking and corresponding helper methods for entities.

### ISSUE THREAD

**Issue by willnationsdev** [dotnet/efcore#28263]
TLDR: DbContexts track unique instances by their primary key and object instance, but *conceptually*, instances are also "unique" based on their unique indexes as well. I would like to see DbContexts adopt helper methods that expand the scope of object comparisons/tracking to include unique indexes so that app code can leverage an entity's model configuration to better automate the creation of quality-of-life methods.

## Scenario

It is very common for me to create entities that have a combination of a primary key, a unique, non-clustered index on some sort of name or guid, and then some set of other properties. So, something like this:

```cs
public class Thing
{
    public int Id { get; set; } // primary key
    [MaxLength(256)]
    public string Name { get; set; } = ""; // has unique index
    public string OtherData { get; set; } = ""; // random other property
}
```

Usually, I will then create helper extension methods for the entity that simplify the process of both 1) finding and/or creating a new instance of the entity, or 2) taking an existing instance of the entity - possibly converted from some other data type - and attempting to upsert the instance into the database *based* on unique index properties rather than on the instance (which will be different) or the primary key (which will usually be the `default` value from such an ad-hoc entity instance).

```cs
public static async Task<Thing> GetOrCreateThingAsync(this DbContext self, string name)
{
    var thing = self.Things.Local.FirstOrDefault(t => t.Name == name);
    thing ??= await self.Things.FirstOrDefaultAsync(t => t.Name == name);
    if (thing is null)
    {
        var thingFactory = self.GetService<ThingFactory>(); // or perhaps an IMyDomainFactory<Thing>
        thing = await thingFactory!.CreateAsync(name);
    }
    return thing;
}

public static async Task<Thing> MergeThingAsync(this DbContext self, Thing thing)
{
    var found = self.Things.Local.FirstOrDefault(t => t.Name == thing.Name);
    found ??= await self.Things.FirstOrDefaultAsync(t => t.Name == thing.Name);
    if (thing is null)
    {
        self.Entry(thing).State = Added;
    }
    else
    {
        // Usually entity-specific "sync" / "merge" logic written here.
        var modified = false;
        if (thing.OtherData != null && found.OtherData != null && !thing.OtherData.Equals(found.OtherData))
        {
            found.OtherData = thing.OtherData;
            modified = true;
        }
        // ...continue comparing and overwriting other non-primary-key, non-unique-index properties. Then...
        if (modified)
        {
            self.Entry(thing).State = Modified;
        }
    }
    await self.SaveChangesAsync();
    return found ?? thing; // calling context can compare object instance to distinguish which code flow executed
}
```

Both of these kinds of operations perform a local search and then a remote search using unique index properties rather than traditional identification methods, and then perform some operation at the end.

## Problems

If I don't do any optimizations, then entities are typically queried by running the predicate over the flat list of entities in the DbSet (which are not indexed by their unique indexes in the server app code side). This becomes wasteful if I do several different lookups of a similar nature on the same DbContext within the same scope.

If I attempt to implement caching, then I have to add a bunch of additional types and logic for the vast majority of my defined entity types. I could create a `record` type or an `IEqualityComparer<T>` for each of the entities, but then I end up effectively having to define the makeup of the unique index in two different places. More than just feeling like a code smell, it also involves its own layer of additional work that must be multiplied for each type of entity *and* each type of unique index ON the entity.

If I ignore the problem altogether, then the alternative is catching and properly handling any duplicate key exceptions that are thrown by the database when the eventual `SaveChangesAsync()` method finally runs. Not only does this throw an exception as code flow though (which I don't like), but it also wastes a trip to the database. All information to know which properties are locally indexable is already supplied *to* the DbContext, and there is already a precedent for `Find` methods that search the local collections before spending effort on a trip to the database and back. It feels like wasted development work on the part of people using the DbContext as a quick-and-easy DAL when it feels as though it wouldn't require many more steps for DbContext to offer the same or similar functionality out-of-the-box.

## Proposed Solution

1. Add logic to DbContext that enables it to (potentially) track entities not only be primary key and object instance, but also by any of the entity's unique indexes.
1. Obviously, having a DbContext track additional information like this would incur a potentially expensive cost. Thus, I would recommend exposing it as an opt-in mechanism on a per-DbContext-instance setting, similar to the `AsNoTracking()` method, i.e. you could add a `WithUniqueIndexTracking()` method.
1. In addition to `Find<T>` and `FindAsync<T>` for `DbSet<T>`, you could also have `FindBy<T, TKey>` and `FindByAsync<T, TKey>` methods that accept either the primary key OR a potential unique index value for the entity. Assuming the `TKey` doesn't match the CLR type of the primary key, it would then be compared against the CLR types of the properties in each defined unique index for a match. If 2 or more unique indexes are constructed from the same combination of CLR types, then an exception would be thrown. If you supplied an anonymous object func or a string list to indicate which unique index you wished to use (like how the index is first defined in `OnModelCreating`), then it would not throw the exception unless the hinted unique index did not refer to anything. There would also need to be a warning in the methods' XML comment docs that explain how they will throw if the DbContext has not opted into the behavior.
1. Add a `FindMatching<T>(T entity)` method + async variant that searches the local and remote collections for any entity that matches any non-default primary key or unique index present on the provided entity. This is more of a utility method for doing multiple `Find` or `FindBy` like operations, but just automatically abstracted on a per-entity-type basis so that the calling code does not need to be aware of each of the different property combinations that could uniquely identify a like object or table row.
1. Add a `MergeNonUniqueProperties<T>(T source, T destination)` method that leverages the existing model information recorded in the DbContext to perform a "smart copy" operation between two entity instances of the same type. Properties that are part of the primary key or any unique indexes and which have non-default values from the source object would be ignored while any other public property would be overwritten, regardless of its value.

**Comment by roji** [comment]
> If I ignore the problem altogether, then the alternative is catching and properly handling any duplicate key exceptions that are thrown by the database when the eventual SaveChangesAsync() method finally runs. Not only does this throw an exception as code flow though (which I don't like), but it also wastes a trip to the database. All information to know which properties are locally indexable is already supplied to the DbContext, and there is already a precedent for Find methods that search the local collections before spending effort on a trip to the database and back.

This seems to ignore the possibility that the database row is inserted concurrently by someone else after the local lookup is done; in other words, a duplicate exception may occur in any case, and calling code must be prepared to handle it (at least in concurrent scenarios). This is the case regardless of whether lookup is done based on the primary key (with Find) or based on some other unique property.

Stepping back, a lot of the above seems to actually be about having full upsert support (#4526); that would provide an API which guarantees that a duplicate key exception wouldn't be thrown. The key point is in order for this to work, the upsert needs to go all the way to the database - it's not enough just to look at EF's local change tracking cache. Once #4526 is done (and it's a highly-requested feature), I think most of the above is no longer needed.

If we set aside the upsert-related requests, one thing that *may* still be useful is "efficient" (index-based) local lookup by a unique, non-PK property. This could simply be a FindBy that accepts a lambda, which the user uses to indicate which property they're interested in (avoiding all the complications of basing the property selection on the generic CLR type, as proposed above in 3 and 4). As you write in 2, having EF maintain internal indexes for additional properties is lots of extra overhead that most people don't want, so this would have to be opt-in, and then we'd want the proposed FindBy to throw if used with a property where that opt-in wasn't done.

However, AFAIK this is the 1st time anyone has ever requested it. Furthermore, in most applications, the number of tracked entities is quite small, so there's no real efficiency advantage with having an index over a unique property - you may as well just scan all of them with with FirstOrDefault, as you've done in GetOrCreateThingAsync above (even databases don't utilize an index and prefer table scans when the number of rows is small).

**Comment by willnationsdev** [comment]
@roji

Thanks for the detailed response!

> that would provide an API which guarantees that a duplicate key exception wouldn't be thrown.

I had heard mention of the AddOrUpdate possibility, but hadn't considered that the implementation could also optionally include a non-PK, unique "arbiter", [as you put it](https://github.com/dotnet/efcore/issues/4526#issuecomment-1159475112). That could work quite well if, say, I just define a static func expression on the entity and then re-use that expression both in the model definition and in any extension methods. For example, `AddIndex(Thing.NameIndexExpression).IsUnique())` + `await ctx.Things.AddOrUpdate(thing).On(Thing.NameIndexExpression).RunAsync()`.

> This could simply be a FindBy that accepts a lambda, which the user uses to indicate which property they're interested in (avoiding all the complications of basing the property selection on the generic CLR type, as proposed above in 3 and 4). As you write in 2, having EF maintain internal indexes for additional properties is lots of extra overhead that most people don't want, so this would have to be opt-in, and then we'd want the proposed FindBy to throw if used with a property where that opt-in wasn't done.

Yeah, this is exactly the kind of API I would like to be able to find built into EF Core. Especially if it could be combined with the "customizable arbiter" `AddOrUpdate` concept in the referenced feature request.

> Furthermore, in most applications, the number of tracked entities is quite small, so there's no real efficiency advantage with having an index over a unique property - you may as well just scan all of them with FirstOrDefault, as you've done in GetOrCreateThingAsync above (even databases don't utilize an index and prefer table scans when the number of rows is small).

Hmmm. I hadn't actually thought about the fact that even databases don't bother with it at those small scales. We do lots of integrations between apps, so usually extracting data from multiple sources, transforming them into a common format for the app, and then loading a view or merging the results into another database. Could be hundreds or thousands of entities. Seemed *just* large enough to warrant creating in-memory lookup tables. I have always just done this intuitively though (felt like a common sense thing) rather than explicitly benchmarking it to confirm a measurable effect.

> However, AFAIK this is the 1st time anyone has ever requested it.

That's honestly pretty surprising to me. If you expose a web API for the database, and the calling code knows to refer to the intended entity by a cross-system name or guid, but doesn't know its internal primary key, it sounds natural to me that it would just use the unique index for identification in place of the internal primary key versus making two API calls to first get the Id by unique index and then request the operation.

To then expose that functionality, I'd have to scale that up to X indexes across Y entities which can quickly snowball into a maintenance hazard (especially if database indexes and the C# DAL/caching code don't define the indexed properties in the same place). I would expect people to be looking for DAL frameworks that can automate/simplify the process of having such alternate lookup mechanisms. But if no one else has that issue, I feel like I must be missing something...

**Comment by roji** [comment]
> Could be hundreds or thousands of entities.

I think this is an important point - if you're loading thousands of entities into EF Core's change tracker, that's likely not a good idea. The EF change tracker was designed to support the unit-of-work pattern (UoW), where some entities are loaded (typically a small number!) in order to then be modified and saved back. It definitely was not meant to be a full-fledged, optimized in-memory database supporting various in-memory lookup patterns.

If your scenario calls for loading a lerge number of entities and then performing lookups over them in-memory, I'd suggest looking at an actual in-memory database implementation (outside of EF).

>> However, AFAIK this is the 1st time anyone has ever requested it.
>
> That's honestly pretty surprising to me.

I agree that the loading entities by a non-PK unique property is something lots of application need. However:

1. In many/most cases it's not necessary to bother with doing a local lookup in EF's change tracker; a UoW typically starts with an empty change tracker, and knows that a given entity hasn't yet been loaded from the database. In other words, in the common case, I don't think there's a need to first check the change tracker - you just load directly from the database.
2. Even where an application does want to check the change tracker before going to the database, it's once again usually not necessary to add an index in order to optimize that operation, because the number of tracked entities doesn't justify that performance-wise. In other words, you can simply use FirstOrDefault over DbSet.Local as you've done in the code above.

To summarize, what you're proposing would only make sense when loading quite a large number of entities, *and* not knowing in a specific place in the UoW whether those entities have already been loaded or not. I don't think that happens very often, and that's why

**Comment by ajcvickers** [comment]
Improving the ability to search for tracked entities is tracked by https://github.com/dotnet/efcore/issues/7391
